src/devices/Tlc1543/Tlc1543.cs

```diff
@@ -0,0 +1,132 @@
+            _shouldDispose = controller == null || shouldDispose;
+            _controller = controller ?? new GpioController();
+
+            _endOfConversion = endOfConversion;
+
+            if (_endOfConversion != -1)
+            {
+                _controller.OpenPin(_endOfConversion, PinMode.InputPullUp);
+            }
+        }
+
+        private static bool IsValidChannel(Channel channel) => channel >= Channel.A0 && channel <= Channel.SelfTestMax;
+
+        /// <summary>
+        /// Reads previous reading and prepares next reading for specified channel
+        /// </summary>
+        /// <param name="channelToCharge">Channel to prepare</param>
+        /// <returns>10 bit value corresponding to relative voltage level on channel</returns>
+        public int ReadPreviousAndChargeChannel(Channel channelToCharge)
+        {
+            if (!IsValidChannel(channelToCharge))
+            {
+                throw new ArgumentOutOfRangeException(nameof(channelToCharge));
+            }
+
+            Span<byte> readBuffer = stackalloc byte[2];
+            Span<byte> writeBuffer = stackalloc byte[2];
+
+            writeBuffer[0] = (byte)((int)channelToCharge << 1);
+            _spiDevice.TransferFullDuplex(writeBuffer, readBuffer);
+
+            int previousReading = ((readBuffer[0] & 0b11111) << 5) | (readBuffer[1] & 0b11111);
+            if (_endOfConversion != -1)
+            {
+                // Wait for ADC to report end of conversion or timeout at max conversion time
+                _controller.WaitForEvent(_endOfConversion, PinEventTypes.Rising, _conversionTime);
+            }
+            else
+            {
+                // Max conversion time (21us) as seen in table on page 10 in TLC1543 documentation
+                DelayHelper.Delay(_conversionTime, false);
+            }
+
+            return previousReading;
+        }
+
+        /// <summary>
+        /// Reads sensor value.
+        /// First cycle: Ask for value on the channel <paramref name="channelNumber"/>.
+        /// Second cycle: Return value from the channel while charging <paramref name="nextChannelToCharge"/>.
+        /// </summary>
+        /// <param name="channelNumber">Channel to be read</param>
+        /// <param name="nextChannelToCharge">Next channel to charge</param>
+        /// <returns>A 10 bit value corresponding to relative voltage level on specified device channel</returns>
+        public int ReadChannel(Channel channelNumber, Channel nextChannelToCharge = Channel.SelfTestHalf)
+        {
+            ReadPreviousAndChargeChannel(channelNumber);
+            return ReadPreviousAndChargeChannel(nextChannelToCharge);
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        public void Dispose()
+        {
+            if (_controller != null)
+            {
+                if (_shouldDispose)
+                {
+                    _controller.Dispose();
+                    _spiDevice.Dispose();

```

Comment:
shouldDispose is for the GpioController only so the I2cDevice should be disposed